packages/yarnpkg-pnp/sources/esm-loader/loader.ts

```diff
@@ -0,0 +1,192 @@
+      }
+    }
+  }
+
+  const result = pnpapi.resolveRequest(specifier, issuer, {
+    conditions: new Set(conditions),
+    // TODO: Handle --experimental-specifier-resolution=node
+    extensions: allowLegacyResolve ? undefined : [],
+  });
+
+  if (!result)
+    throw new Error(`Resolution failed`);
+
+  return {
+    url: pathToFileURL(result).href,
+  };
+}
+
+const realModules = new Set<string>();
+
+export async function getFormat(
+  resolved: string,
+  context: any,
+  defaultGetFormat: any
+) {
+  const parsedURL = new URL(resolved);
+  if (parsedURL.protocol !== `file:`)
+    return defaultGetFormat(resolved, context, defaultGetFormat);
+
+  const ext = path.extname(parsedURL.pathname);
+  switch (ext) {
+    case `.mjs`: {
+      realModules.add(fileURLToPath(resolved));
+      return {
+        format: `module`,
+      };
+    }
+    case `.cjs`: {
+      return {
+        format: `module`,
+      };
+    }
+    case `.json`: {
+      // TODO: Enable if --experimental-json-modules is present
+      throw new Error(
+        `Unknown file extension ".json" for ${fileURLToPath(resolved)}`
+      );
+      return {
+        format: `module`,
+      };
+    }
+    case `.js`: {
+      const filePath = fileURLToPath(resolved);
+      const pkg = nodeUtils.readPackageScope(filePath);
+      if (pkg) {
+        let moduleType = pkg.data.type ?? `commonjs`;
+        if (moduleType === `commonjs`) moduleType = `module`;
+        else realModules.add(filePath);
+
+        return {
+          format: moduleType,
+        };
+      }
+    }
+  }
+
+  return defaultGetFormat(resolved, context, defaultGetFormat);
+}
+
+let parserInit: Promise<void> | null = init().then(() => {
+  parserInit = null;
+});
+
+async function parseExports(filePath: string) {
+  const {exports} = parse(await fs.promises.readFile(filePath, `utf8`));
+
+  return new Set(exports);
+}
+
+export async function getSource(
+  urlString: string,
+  context: any,
+  defaultGetSource: any
+) {
+  const url = new URL(urlString);
+  if (url.protocol !== `file:`)
+    return defaultGetSource(url, context, defaultGetSource);
+
+  urlString = fileURLToPath(urlString);
+
+  if (realModules.has(urlString)) {
+    return {
+      source: await fs.promises.readFile(urlString, `utf8`),
+    };
+  }
+
+  if (parserInit !== null) await parserInit;
+
+  const exports = await parseExports(urlString);
+
+  let exportStrings = `export default cjs\n`;
+  for (const exportName of exports) {
+    if (exportName !== `default`) {
+      exportStrings += `const __${exportName} = cjs['${exportName}'];\n export { __${exportName} as ${exportName} }\n`;
+    }
+  }
+
+  const fakeModulePath = path.join(path.dirname(urlString), `noop.js`);
+
+  const code = `
+  import {createRequire} from 'module';
+  const require = createRequire('${fakeModulePath.replace(/\\/g, `/`)}');
+  const cjs = require('${urlString.replace(/\\/g, `/`)}');
+
+  ${exportStrings}
+  `;
```

Comment:
I don't understand why we need to generate code here

Reply:
Pretty much nothing in the ESM code path in Node uses the patched `fs` so we have to implement a lot of hacks, this is one of them, the loader is only allowed to return source code so it's reimplementing the internal translator.
https://github.com/nodejs/node/blob/b7136827f5d9f4f65bd0a60e75c4a9531cf4312a/lib/internal/modules/esm/translators.js#L180-L221